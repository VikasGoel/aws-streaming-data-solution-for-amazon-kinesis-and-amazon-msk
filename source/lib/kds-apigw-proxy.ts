/*********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as iam from '@aws-cdk/aws-iam';
import * as cwlogs from '@aws-cdk/aws-logs';
import * as kinesis from '@aws-cdk/aws-kinesis';

export interface ProxyApiProps {
    readonly stream: kinesis.IStream;
    readonly accessLogsRetention: cwlogs.RetentionDays;
    readonly throttlingRateLimit: number;
    readonly throttlingBurstLimit: number;
}

interface IntegrationProps {
    readonly action: string;
    readonly path: string;
    readonly requestTemplate: string;
    readonly role: iam.Role;
    readonly model: apigw.IModel;
    readonly validator: apigw.IRequestValidator;
}

export class ProxyApi extends cdk.Construct {
    public readonly Api: apigw.RestApi;

    constructor(scope: cdk.Construct, id: string, props: ProxyApiProps) {
        super(scope, id);

        if (!cdk.Token.isUnresolved(props.throttlingRateLimit) && props.throttlingRateLimit <= 0) {
            throw new Error('throttlingRateLimit must be a positive number');
        }

        if (!cdk.Token.isUnresolved(props.throttlingBurstLimit) && props.throttlingBurstLimit < 0) {
            throw new Error('throttlingBurstLimit must be a non-negative number');
        }

        const logGroup = new cwlogs.LogGroup(this, 'AccessLogs', {
            retention: props.accessLogsRetention,
            removalPolicy: cdk.RemovalPolicy.RETAIN
        });

        const apiGatewayRole = new iam.Role(this, 'ApiGatewayRole', {
            assumedBy: new iam.ServicePrincipal('apigateway.amazonaws.com')
        });

        props.stream.grantWrite(apiGatewayRole);

        this.Api = new apigw.RestApi(this, 'Api', {
            restApiName: `${cdk.Aws.STACK_NAME}-kinesis-proxy`,
            cloudWatchRole: true,
            deployOptions: {
                methodOptions: {
                    '/*/*': {
                        throttlingRateLimit: props.throttlingRateLimit,
                        throttlingBurstLimit: props.throttlingBurstLimit
                    }
                },
                accessLogDestination: new apigw.LogGroupLogDestination(logGroup),
                accessLogFormat: apigw.AccessLogFormat.clf()
            }
        });

        const validator = this.Api.addRequestValidator('Validator', {
            requestValidatorName: 'default-validator',
            validateRequestBody: true
        });

        this.addIntegration({
            action: 'PutRecord',
            path: 'record',
            model: this.getPutRecordModel(),
            requestTemplate: this.getPutRecordRequestTemplate(props.stream.streamName),
            role: apiGatewayRole,
            validator
        });

        this.addIntegration({
            action: 'PutRecords',
            path: 'records',
            model: this.getPutRecordsModel(),
            requestTemplate: this.getPutRecordsRequestTemplate(props.stream.streamName),
            role: apiGatewayRole,
            validator
        });

        this.addCfnNagSuppressions();
    }

    private getPutRecordRequestTemplate(streamName: string): string {
        return `{ "StreamName": "${streamName}", "Data": "$util.base64Encode($input.json('$.data'))", "PartitionKey": "$input.path('$.partitionKey')" }`;
    }

    private getPutRecordModel(): apigw.IModel {
        return this.Api.addModel('PutRecordModel', {
            contentType: 'application/json',
            modelName: 'PutRecordModel',
            description: 'PutRecord proxy single-record payload',
            schema: {
                schema: apigw.JsonSchemaVersion.DRAFT4,
                title: 'PutRecord proxy single-record payload',
                type: apigw.JsonSchemaType.OBJECT,
                required: ['data', 'partitionKey'],
                properties: {
                    data: { type: apigw.JsonSchemaType.STRING },
                    partitionKey: { type: apigw.JsonSchemaType.STRING }
                }
            }
        });
    }

    private getPutRecordsRequestTemplate(streamName: string): string {
        return `{ "StreamName": "${streamName}", "Records": [ #foreach($elem in $input.path('$.records')) { "Data": "$util.base64Encode($elem.data)", "PartitionKey": "$elem.partitionKey"}#if($foreach.hasNext),#end #end ] }`;
    }

    private getPutRecordsModel(): apigw.IModel {
        return this.Api.addModel('PutRecordsModel', {
            contentType: 'application/json',
            modelName: 'PutRecordsModel',
            description: 'PutRecords proxy payload data',
            schema: {
                schema: apigw.JsonSchemaVersion.DRAFT4,
                title: 'PutRecords proxy payload data',
                type: apigw.JsonSchemaType.OBJECT,
                required: ['records'],
                properties: {
                    records: {
                        type: apigw.JsonSchemaType.ARRAY,
                        items: {
                            type: apigw.JsonSchemaType.OBJECT,
                            required: ['data', 'partitionKey'],
                            properties: {
                                data: { type: apigw.JsonSchemaType.STRING },
                                partitionKey: { type: apigw.JsonSchemaType.STRING }
                            }
                        }
                    }
                }
            }
        });
    }

    private addIntegration(props: IntegrationProps)  {
        const integration = new apigw.AwsIntegration({
            service: 'kinesis',
            integrationHttpMethod: 'POST',
            action: props.action,
            options: {
                credentialsRole: props.role,
                passthroughBehavior: apigw.PassthroughBehavior.NEVER,
                requestParameters: {
                    'integration.request.header.Content-Type': "'x-amz-json-1.1'"
                },
                requestTemplates: {
                    'application/json': props.requestTemplate
                },
                integrationResponses: [
                    {
                        statusCode: '200'
                    },
                    {
                        statusCode: '500',
                        selectionPattern: '5\\d{2}'
                    }
                ]
            }
        });

        const resource = this.Api.root.addResource(props.path, {
            defaultCorsPreflightOptions: {
                allowOrigins: apigw.Cors.ALL_ORIGINS,
                allowMethods: ['PUT']
            }
        });

        resource.addMethod('PUT', integration, {
            authorizationType: apigw.AuthorizationType.IAM,
            requestModels: {
                'application/json': props.model
            },
            requestValidator: props.validator,
            methodResponses: [
                {
                    statusCode: '200',
                    responseParameters: { 'method.response.header.Content-Type': true }
                },
                {
                    statusCode: '500',
                    responseParameters: { 'method.response.header.Content-Type': true }
                }
            ]
        });
    }

    private addCfnNagSuppressions() {
        const cfnDeployment = this.Api.latestDeployment?.node.defaultChild as apigw.CfnDeployment;
        cfnDeployment.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W68',
                    reason: 'Default usage plan can be used for this API'
                }]
            }
        };

        const cfnStage = this.Api.deploymentStage.node.defaultChild as apigw.CfnStage;
        cfnStage.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W64',
                    reason: 'Default usage plan can be used for this API'
                }]
            }
        };
    }
}
