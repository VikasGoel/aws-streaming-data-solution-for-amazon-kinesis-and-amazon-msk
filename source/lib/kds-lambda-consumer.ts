/*********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as kinesis from '@aws-cdk/aws-kinesis';
import * as sqs from '@aws-cdk/aws-sqs';
import * as cw from '@aws-cdk/aws-cloudwatch';

import { KinesisEventSource, SqsDlq } from '@aws-cdk/aws-lambda-event-sources';
import { ExecutionRole } from './lambda-role-cloudwatch';

export interface LambdaConsumerProps {
    readonly stream: kinesis.IStream;
    readonly batchSize: number;
    readonly retryAttempts: number;
    readonly parallelizationFactor: number;
    readonly timeout: cdk.Duration;
}

export class LambdaConsumer extends cdk.Construct {
    public readonly Function: lambda.IFunction;

    private MIN_TIMEOUT_SECONDS: number = 1;
    private MAX_TIMEOUT_SECONDS: number = 900;

    constructor(scope: cdk.Construct, id: string, props: LambdaConsumerProps) {
        super(scope, id);

        const timeoutSeconds = props.timeout.toSeconds();
        if (timeoutSeconds < this.MIN_TIMEOUT_SECONDS || timeoutSeconds > this.MAX_TIMEOUT_SECONDS) {
            throw new Error(`timeout must be a value between ${this.MIN_TIMEOUT_SECONDS} and ${this.MAX_TIMEOUT_SECONDS} seconds`);
        }

        const dlq = this.createDeadLetterQueue();
        const executionRole = new ExecutionRole(this, 'Role');

        this.Function = new lambda.Function(this, 'Consumer', {
            runtime: lambda.Runtime.NODEJS_12_X,
            handler: 'index.handler',
            role: executionRole.Role,
            code: lambda.Code.fromAsset('lambda/kds-lambda-consumer'),
            timeout: props.timeout
        });

        this.Function.addEventSource(new KinesisEventSource(props.stream, {
            startingPosition: lambda.StartingPosition.LATEST,
            batchSize: props.batchSize,
            retryAttempts: props.retryAttempts,
            parallelizationFactor: props.parallelizationFactor,
            bisectBatchOnError: true,
            onFailure: new SqsDlq(dlq)
        }));
    }

    private createDeadLetterQueue(): sqs.Queue {
        const dlq = new sqs.Queue(this, 'DLQ', {
            encryption: sqs.QueueEncryption.KMS_MANAGED,
            dataKeyReuse: cdk.Duration.minutes(30)
        });

        new cw.Alarm(this, 'DLQDepthAlarm', {
            metric: new cw.Metric({
                namespace: 'AWS/SQS',
                metricName: 'ApproximateNumberOfMessagesVisible',
                dimensions: { 'QueueName': dlq.queueName },
                statistic: 'Sum',
                period: cdk.Duration.minutes(5)
            }),
            threshold: 1,
            comparisonOperator: cw.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
            evaluationPeriods: 1,
            treatMissingData: cw.TreatMissingData.NOT_BREACHING
        });

        return dlq;
    }
}