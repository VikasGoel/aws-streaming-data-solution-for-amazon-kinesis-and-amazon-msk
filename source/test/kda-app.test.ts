/*********************************************************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                           *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as logs from '@aws-cdk/aws-logs';
import * as kinesis from '@aws-cdk/aws-kinesis';
import * as s3 from '@aws-cdk/aws-s3';
import { expect as expectCDK, haveResource, haveResourceLike, ResourcePart, SynthUtils } from '@aws-cdk/assert';

import { FlinkApplication } from '../lib/kda-flink-application';

describe('successful scenarios', () => {
    let stack: cdk.Stack;

    beforeAll(() => {
        const app = new cdk.App();
        stack = new cdk.Stack(app, 'TestStack');

        new FlinkApplication(stack, 'TestApplication', {
            inputStream: new kinesis.Stream(stack, 'TestStream'),
            outputBucket: new s3.Bucket(stack, 'TestBucket'),

            logsRetentionDays: logs.RetentionDays.ONE_WEEK,
            logLevel: 'INFO',
            metricsLevel: 'APPLICATION',

            codeBucketArn: 'arn:aws:s3:::some-bucket',
            codeFileKey: 'some-key.zip',

            enableAutoScaling: 'true',
            enableSnapshots: 'true',

            subnetIds: ['subnet-a', 'subnet-b'],
            securityGroupIds: ['sg-123']
        });
    });

    test('creates a KDA flink application', () => {
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    });

    test('configures logging for the KDA application', () => {
        expectCDK(stack).to(haveResource('AWS::Logs::LogGroup', {
            RetentionInDays: 7
        }));

        expectCDK(stack).to(haveResource('AWS::Logs::LogStream'));
        expectCDK(stack).to(haveResource('AWS::KinesisAnalyticsV2::ApplicationCloudWatchLoggingOption'));
    });

    test('creates an IAM role for KDA', () => {
        expectCDK(stack).to(haveResource('AWS::IAM::Role', {
            AssumeRolePolicyDocument: {
                Statement: [{
                    Action: 'sts:AssumeRole',
                    Effect: 'Allow',
                    Principal: {
                        Service: 'kinesisanalytics.amazonaws.com'
                    }
                }],
                Version: '2012-10-17'
            }
        }));

        expectCDK(stack).to(haveResourceLike('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: [
                    {
                        Action: [
                            'kinesis:DescribeStreamSummary',
                            'kinesis:GetRecords',
                            'kinesis:GetShardIterator',
                            'kinesis:ListShards',
                            'kinesis:SubscribeToShard'
                        ],
                        Effect: 'Allow'
                    },
                    {
                        Action: [
                            's3:GetObject*',
                            's3:GetBucket*',
                            's3:List*',
                            's3:DeleteObject*',
                            's3:PutObject*',
                            's3:Abort*'
                        ],
                        Effect: 'Allow'
                    }
                ]
            }
        }));

        expectCDK(stack).to(haveResourceLike('AWS::IAM::Policy', {
            PolicyDocument: {
                Statement: [
                    {
                        Action: [
                            'ec2:CreateNetworkInterface',
                            'ec2:DescribeNetworkInterfaces',
                            'ec2:DescribeVpcs',
                            'ec2:DeleteNetworkInterface',
                            'ec2:DescribeDhcpOptions',
                            'ec2:DescribeSubnets',
                            'ec2:DescribeSecurityGroups'
                        ],
                        Effect: 'Allow'
                    },
                    {
                        Action: 'ec2:CreateNetworkInterfacePermission',
                        Effect: 'Allow'
                    }
                ]
            }
        }));
    });

    test('adds cfn_nag suppressions', () => {
        expectCDK(stack).to(haveResource('AWS::IAM::Policy', {
            Metadata: {
                cfn_nag: {
                    rules_to_suppress: [{
                        id: 'W12',
                        reason: 'Actions do not support resource level permissions'
                    }]
                }
            }
        }, ResourcePart.CompleteDefinition));
    });
});

describe('validation tests', () => {
    let stack: cdk.Stack;
    let fakeStream: kinesis.IStream;
    let fakeBucket: s3.IBucket;

    beforeEach(() => {
        const app = new cdk.App();
        stack = new cdk.Stack(app, 'TestStack');
        fakeStream = new kinesis.Stream(stack, 'TestStream');
        fakeBucket = new s3.Bucket(stack, 'TestBucket');
    });

    test('invalid log level', () => {
        expect(() => new FlinkApplication(stack, 'TestApplication', {
            inputStream: fakeStream,
            outputBucket: fakeBucket,

            logsRetentionDays: logs.RetentionDays.ONE_WEEK,
            logLevel: 'FOO',
            metricsLevel: 'APPLICATION',

            codeBucketArn: 'arn:aws:s3:::some-bucket',
            codeFileKey: 'some-key.zip',

            enableAutoScaling: 'true',
            enableSnapshots: 'true'
        })).toThrowError(/Unknown log level: FOO/);
    });

    test('invalid metric level', () => {
        expect(() => new FlinkApplication(stack, 'TestApplication', {
            inputStream: fakeStream,
            outputBucket: fakeBucket,

            logsRetentionDays: logs.RetentionDays.ONE_WEEK,
            logLevel: 'INFO',
            metricsLevel: 'BAR',

            codeBucketArn: 'arn:aws:s3:::some-bucket',
            codeFileKey: 'some-key.zip',

            enableAutoScaling: 'true',
            enableSnapshots: 'true'
        })).toThrowError(/Unknown metrics level: BAR/);
    });
});
