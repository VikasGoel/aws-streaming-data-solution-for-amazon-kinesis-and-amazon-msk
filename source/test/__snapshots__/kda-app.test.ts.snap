// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successful scenarios creates a KDA flink application 1`] = `
Object {
  "Conditions": Object {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": Object {
      "Fn::Or": Array [
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        Object {
          "Fn::Equals": Array [
            Object {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
    "TestApplicationEnableAutoScaling3998F3DF": Object {
      "Fn::Equals": Array [
        "true",
        "true",
      ],
    },
    "TestApplicationEnableSnapshotsB729B643": Object {
      "Fn::Equals": Array [
        "true",
        "true",
      ],
    },
  },
  "Parameters": Object {
    "AssetParameters6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29ArtifactHashA929B04E": Object {
      "Description": "Artifact hash for asset \\"6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29\\"",
      "Type": "String",
    },
    "AssetParameters6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29S3Bucket6CBC124E": Object {
      "Description": "S3 bucket for asset \\"6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29\\"",
      "Type": "String",
    },
    "AssetParameters6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29S3VersionKey35BC6A36": Object {
      "Description": "S3 key for asset version \\"6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestApplicationAppRoleDefaultPolicy4584665A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestStreamE6F40222",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "TestBucket560B80BC",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "TestBucket560B80BC",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestApplicationAppRoleDefaultPolicy4584665A",
        "Roles": Array [
          Object {
            "Ref": "TestApplicationAppRoleF109DF67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestApplicationAppRoleF109DF67": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "kinesisanalytics.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "TestApplicationCodePolicy8796F599": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObjectVersion",
                "s3:GetObject",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::some-bucket/some-key.zip",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestApplicationCodePolicy8796F599",
        "Roles": Array [
          Object {
            "Ref": "TestApplicationAppRoleF109DF67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestApplicationCustomResource71741B78": Object {
      "DependsOn": Array [
        "TestApplicationCustomResourceRoleC6612C60",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29S3Bucket6CBC124E",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29S3VersionKey35BC6A36",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters6467ddf591ee8d331757e026c674b30fde5ccf2b7239abd2a53645055b04fd29S3VersionKey35BC6A36",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lambda_function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestApplicationCustomResourceRoleC6612C60",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestApplicationCustomResourceRoleC6612C60": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchLogsPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "kinesisanalytics:AddApplicationVpcConfiguration",
                    "kinesisanalytics:DeleteApplicationVpcConfiguration",
                    "kinesisanalytics:DescribeApplication",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":kinesisanalytics:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":application/",
                        Object {
                          "Ref": "TestApplicationFF75924F",
                        },
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "VpcConfigPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestApplicationFF75924F": Object {
      "Properties": Object {
        "ApplicationConfiguration": Object {
          "ApplicationCodeConfiguration": Object {
            "CodeContent": Object {
              "S3ContentLocation": Object {
                "BucketARN": "arn:aws:s3:::some-bucket",
                "FileKey": "some-key.zip",
              },
            },
            "CodeContentType": "ZIPFILE",
          },
          "ApplicationSnapshotConfiguration": Object {
            "SnapshotsEnabled": Object {
              "Fn::If": Array [
                "TestApplicationEnableSnapshotsB729B643",
                true,
                false,
              ],
            },
          },
          "EnvironmentProperties": Object {
            "PropertyGroups": Array [
              Object {
                "PropertyGroupId": "FlinkApplicationProperties",
                "PropertyMap": Object {
                  "InputStreamName": Object {
                    "Ref": "TestStreamE6F40222",
                  },
                  "OutputBucketName": Object {
                    "Ref": "TestBucket560B80BC",
                  },
                  "Region": Object {
                    "Ref": "AWS::Region",
                  },
                },
              },
            ],
          },
          "FlinkApplicationConfiguration": Object {
            "CheckpointConfiguration": Object {
              "ConfigurationType": "DEFAULT",
            },
            "MonitoringConfiguration": Object {
              "ConfigurationType": "CUSTOM",
              "LogLevel": "INFO",
              "MetricsLevel": "APPLICATION",
            },
            "ParallelismConfiguration": Object {
              "AutoScalingEnabled": Object {
                "Fn::If": Array [
                  "TestApplicationEnableAutoScaling3998F3DF",
                  true,
                  false,
                ],
              },
              "ConfigurationType": "CUSTOM",
            },
          },
        },
        "RuntimeEnvironment": "FLINK-1_8",
        "ServiceExecutionRole": Object {
          "Fn::GetAtt": Array [
            "TestApplicationAppRoleF109DF67",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KinesisAnalyticsV2::Application",
    },
    "TestApplicationLogGroup970A3DC7": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "TestApplicationLogStream41304A53": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "LogGroupName": Object {
          "Ref": "TestApplicationLogGroup970A3DC7",
        },
      },
      "Type": "AWS::Logs::LogStream",
      "UpdateReplacePolicy": "Retain",
    },
    "TestApplicationLoggingB343A6A7": Object {
      "Properties": Object {
        "ApplicationName": Object {
          "Ref": "TestApplicationFF75924F",
        },
        "CloudWatchLoggingOption": Object {
          "LogStreamARN": Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":logs:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":",
                Object {
                  "Ref": "AWS::AccountId",
                },
                ":log-group:",
                Object {
                  "Ref": "TestApplicationLogGroup970A3DC7",
                },
                ":log-stream:",
                Object {
                  "Ref": "TestApplicationLogStream41304A53",
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::KinesisAnalyticsV2::ApplicationCloudWatchLoggingOption",
    },
    "TestApplicationLogsPolicyAFC3EE65": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:DescribeLogGroups",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "logs:DescribeLogStreams",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TestApplicationLogGroup970A3DC7",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestApplicationLogsPolicyAFC3EE65",
        "Roles": Array [
          Object {
            "Ref": "TestApplicationAppRoleF109DF67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestApplicationVpcConfiguration6BA87B39": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ApplicationName": Object {
          "Ref": "TestApplicationFF75924F",
        },
        "SecurityGroupIds": Array [
          "sg-123",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "TestApplicationCustomResource71741B78",
            "Arn",
          ],
        },
        "SubnetIds": Array [
          "subnet-a",
          "subnet-b",
        ],
      },
      "Type": "Custom::VpcConfiguration",
      "UpdateReplacePolicy": "Delete",
    },
    "TestApplicationVpcPolicy7E12ABFD": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W12",
              "reason": "Actions do not support resource level permissions",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ec2:CreateNetworkInterface",
                "ec2:DescribeNetworkInterfaces",
                "ec2:DescribeVpcs",
                "ec2:DeleteNetworkInterface",
                "ec2:DescribeDhcpOptions",
                "ec2:DescribeSubnets",
                "ec2:DescribeSecurityGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "ec2:CreateNetworkInterfacePermission",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ec2:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":network-interface/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "TestApplicationVpcPolicy7E12ABFD",
        "Roles": Array [
          Object {
            "Ref": "TestApplicationAppRoleF109DF67",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestBucket560B80BC": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestStreamE6F40222": Object {
      "Properties": Object {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": Object {
          "Fn::If": Array [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            Object {
              "Ref": "AWS::NoValue",
            },
            Object {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
